src/libopensc/card-piv.c

```diff
@@ -507,209 +843,1967 @@ static struct sc_card_driver piv_drv = {
+			break;
+		case 0x87: /* GENERAL AUTHENTICATE */
+			break;
+		default: /* just issue the plain apdu */
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_SM_NOT_APPLIED);
+	}
 
-		/* TODO: -DEE Could add key to cache so could use engine to generate key,
-		 * and sign req in single operation */
-		r = 0;
+	*sm_apdu = calloc(1, sizeof(sc_apdu_t));
+	if (*sm_apdu == NULL) {
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+
+	r = piv_encode_apdu(card, plain, *sm_apdu);
+	if (r < 0 && *sm_apdu) {
+		piv_free_sm_apdu(card, NULL, sm_apdu);
 	}
 
-err:
 	LOG_FUNC_RETURN(card->ctx, r);
 }
 
 
-static int piv_select_aid(sc_card_t* card, u8* aid, size_t aidlen, u8* response, size_t *responselen)
+/* ASN1 callback to save address and len of the object */
+static int piv_get_asn1_obj(sc_context_t *ctx, void *arg,  const u8 *obj, size_t len, int depth)
 {
-	sc_apdu_t apdu;
-	int r;
+	struct sc_lv_data *al = arg;
 
-	LOG_FUNC_CALLED(card->ctx);
+	if (!arg)
+		return SC_ERROR_INTERNAL;
+
+	al->value = (u8 *)obj;
+	al->len = len;
+	return SC_SUCCESS;
+}
+
+static int piv_decode_apdu(sc_card_t *card, sc_apdu_t *plain, sc_apdu_t *sm_apdu)
+{
+	int r = SC_SUCCESS;
+	int i;
+	piv_private_data_t * priv = PIV_DATA(card);
+	cipher_suite_t *cs = priv->cs;
+	struct sc_lv_data ee = {NULL, 0};
+	struct sc_lv_data status = {NULL, 0};
+	struct sc_lv_data rmac8 = {NULL, 0};
+	u8 zeros[16] = {0};
+	u8 IV[16];
+	u8 *p;
+	int outl;
+	int outli;
+	int outll;
+	int outdl;
+	u8 lastb[16];
+	u8 discard[8];
+	u8 *q = NULL;
+	int inlen;
+	int macdatalen;
+
+	size_t MCVlen = 16;
+	size_t R_MCVlen = 0;
+
+	EVP_CIPHER_CTX *ed_ctx = NULL;
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+	CMAC_CTX *cmac_ctx  = NULL;
+#else
+	EVP_MAC *mac = NULL;
+	EVP_MAC_CTX *cmac_ctx = NULL;
+	OSSL_PARAM cmac_params[2];
+	size_t cmac_params_n = 0;
+#endif
+
+	struct sc_asn1_entry asn1_sm_response[C_ASN1_PIV_SM_RESPONSE_SIZE];
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	sc_copy_asn1_entry(c_asn1_sm_response, asn1_sm_response);
+
+	sc_format_asn1_entry(asn1_sm_response + 0, piv_get_asn1_obj, &ee, 0);
+	sc_format_asn1_entry(asn1_sm_response + 1, piv_get_asn1_obj, &status, 0);
+	sc_format_asn1_entry(asn1_sm_response + 2, piv_get_asn1_obj, &rmac8, 0);
+
+	r = sc_asn1_decode(card->ctx, asn1_sm_response, sm_apdu->resp, sm_apdu->resplen, NULL, NULL);
+
+	if (r < 0) {
+		sc_log(card->ctx,"SM decode failed");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	if (asn1_sm_response[0].flags & SC_ASN1_PRESENT  /* optional */
+			&& ( ee.value == NULL || ee.len <= 2)) {
+		sc_log(card->ctx,"SM BER-TLV not valid");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	/* if no data returned clear plain resplen */
+	if (!(asn1_sm_response[0].flags & SC_ASN1_PRESENT)) {
+		plain->resplen = 0;
+	}
+
+	if ((asn1_sm_response[1].flags & SC_ASN1_PRESENT) == 0
+			|| (asn1_sm_response[2].flags & SC_ASN1_PRESENT) == 0) {
+		sc_log(card->ctx,"SM missing status or R-MAC");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
```

Comment:
Typically, you want to check authentication first, before handling any decryption (including setting `plain->resplen = 0;`). Please consider checking the MAC before processing the encrypted data.

Reply:
I believe it is checking the MAC before decrypting the response. 
Line 1227         r = sc_asn1_decode(card->ctx, asn1_sm_response, sm_apdu->resp, sm_apdu->resplen, NULL, NULL);

where "ee",  status and R-MAC are decoded.  "ee" is the TLV for the encrypted data. If "ee" is present size must be > 2 i.e. 0ne byte tag, one byte length one byte padding indicator.  The Value is not examined here.  If "ee" is not present then `plain->resplen = 0;` If "ee" is present `plain->resplen` is set in line 1402 `plain->resplen = inlen + i;`

lines 1247 - 1261 check the sizes of the status and R-MAC

Lines 1280 - 1307 calculate the R-MAC and compare to what the card says is the R-MAC.

Lines 1309 - 1385 decrypt the "ee" value and check the trailing padding is correct.

Reply:
I was referring to [line 1244](https://github.com/OpenSC/OpenSC/blob/0bf36bbb3fcfd02469cd3f5d3347d210602b0064/src/libopensc/card-piv.c#L1244), which may be better moved after the check of the MAC